pkg/util/metrics_helper.go

```diff
@@ -0,0 +1,311 @@
+package util
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+// MetricFamiliesPerUser is a collection of metrics gathered via calling Gatherer.Gather() method on different
+// gatherers, one per user.
+// First key = userID, second key = metric name.
+// Value = slice of gathered values with the same metric name.
+type MetricFamiliesPerUser map[string]map[string]*dto.MetricFamily
+
+func BuildMetricFamiliesPerUserFromUserRegistries(regs map[string]*prometheus.Registry) MetricFamiliesPerUser {
+	data := MetricFamiliesPerUser{}
+	for userID, r := range regs {
+		m, err := r.Gather()
+		if err == nil {
+			err = data.AddGatheredDataForUser(userID, m)
+		}
+
+		if err != nil {
+			level.Warn(Logger).Log("msg", "failed to gather metrics from TSDB shipper", "user", userID, "err", err)
+			continue
+		}
+	}
+	return data
+}
+
+// AddGatheredDataForUser adds user-specific output of Gatherer.Gather method.
+// Gatherer.Gather specifies that there metric families are uniquely named, and we use that fact here.
+// If they are not, this method returns error.
+func (d MetricFamiliesPerUser) AddGatheredDataForUser(userID string, metrics []*dto.MetricFamily) error {
+	// Keeping map of metric name to its family makes it easier to do searches later.
+	perMetricName := map[string]*dto.MetricFamily{}
+
+	for _, m := range metrics {
+		name := m.GetName()
+		// these errors should never happen when passing Gatherer.Gather() output.
+		if name == "" {
+			return errors.New("empty name for metric family")
+		}
+		if perMetricName[name] != nil {
+			return fmt.Errorf("non-unique name for metric family: %q", name)
+		}
+
+		perMetricName[name] = m
+	}
+
+	d[userID] = perMetricName
+	return nil
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCounters(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string) {
+	result := float64(0)
+	for _, perMetric := range d {
+		result += sum(perMetric[counter], counterValue)
```

Comment:
If for any reason `perMetric[counter]` is `nil`, the `sum()` wil panic. Would be also great to cover this edge case with a test.

_Same comment applies to other `Send...()` functions._

Reply:
That's not true. range of nil slice doesn't panic.

Reply:
But `range mf.Metric {` will panic if `mf` is `nil`

Reply:
I don't use `range mf.Metric`, but `range mf.GetMetric()`, which will not panic, even for `mf` being `nil`.

Reply:
Ah, sorry, you're right... I use `mf.Metric` in `sum`... needs to be changed.

Reply:
Fixed, and added test for sum function.